File: Arrays/02-second_smallest/Python/optimal.py
from typing import List

# Optimal Approach
# ---------------------
# INTUITION - Take the smallest and the second smallest variables and update both of them as per conditions
# TIME COMPLEXITY - O(N)
# SPACE COMPLEXITY - O(1)

def find_second_smallest_optimal( array : List[int] ) -> int :
    smallest = float("inf")
    second_smallest = float("inf")
    for num in array:
        if num < smallest: 
            second_smallest = smallest
            smallest = num
        elif num < second_smallest and num > smallest:
            second_smallest = num
    
    return second_smallest if second_smallest != float("inf") else None

# CHECK
arr = [2]
print(f"Second smallest in {arr} : {find_second_smallest_optimal(arr)}")

File: Arrays/02-second_smallest/C++/optimal.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

/*
Optimal Approach
---------------------
INTUITION - Take the smallest and the second smallest variables and update both of them as per conditions
TIME COMPLEXITY - O(N)
SPACE COMPLEXITY - O(1)
*/

int find_second_smallest_optimal( vector<int> array ){
    int smallest = INT_MAX;
    int second_smallest = INT_MAX;
    for (auto num: array){
        if (num < smallest){
            second_smallest = smallest;
            smallest = num;
        }
        else if (num < second_smallest){
            second_smallest = num;
        }
    }   
    return (second_smallest == INT_MAX) ? NULL : second_smallest;
}

int main() {
    vector<int> vec = {0};
    cout << "Second smallest : " << find_second_smallest_optimal(vec) << endl;
    return 0;
}